src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -1177,50 +1196,37 @@ void RegisterXdsResolver(CoreConfiguration::Builder* builder) {
       std::make_unique<XdsResolverFactory>());
 }
 
-void XdsClusterMap::Put(RefCountedPtr<ClusterState> state) {
-  absl::string_view cluster_name = state->cluster();
-  clusters_.emplace(cluster_name, std::move(state));
-}
-
-absl::optional<std::pair<absl::string_view, RefCountedPtr<ClusterState>>>
-XdsClusterMap::Find(absl::string_view name) const {
+RefCountedPtr<ClusterState> XdsClusterMap::Find(absl::string_view name) const {
   auto it = clusters_.find(name);
   if (it == clusters_.end()) {
-    return absl::nullopt;
+    return nullptr;
   }
-  return *it;
+  return it->second;
 }
 
 ClusterState::ClusterState(RefCountedPtr<XdsResolver> resolver,
-                           const std::string& cluster_name)
-    : resolver_(std::move(resolver)),
-      it_(resolver_->Emplace(cluster_name, WeakRef())) {}
+                           absl::string_view cluster_name)
+    : resolver_(std::move(resolver)), cluster_name_(cluster_name) {}
 
 void ClusterState::Orphan() {
-  auto* resolver = resolver_.get();
-  resolver->Run([resolver = std::move(
-                     resolver_)]() { resolver->MaybeRemoveUnusedClusters(); },
-                DEBUG_LOCATION);
+  resolver_->MaybeRemoveUnusedClustersInWorkSerializer();
 }
 
-XdsClusterLbDataAttribute::XdsClusterLbDataAttribute(
+XdsClusterDataAttribute::XdsClusterDataAttribute(
     RefCountedPtr<XdsClusterMap> cluster_map)
     : cluster_map_(std::move(cluster_map)) {}
 
-bool XdsClusterLbDataAttribute::LockClusterConfig(
+RefCountedPtr<ClusterState> XdsClusterDataAttribute::LockAndGetClusterConfig(
     absl::string_view cluster_name) {
   if (cluster_map_ == nullptr) {
-    gpr_log(GPR_ERROR, "Cluster map already released");
-    return false;
+    return nullptr;
   }
-  auto cluster_info = cluster_map_->Find(cluster_name);
-  if (cluster_info.has_value()) {
-    locked_cluster_config_ = cluster_info->second;
-    cluster_map_.reset();
-    return true;
+  auto cluster = cluster_map_->Find(cluster_name);
+  if (cluster == nullptr) {
```

Comment:
This should never happen.  But if it did, I think we still want to unref `cluster_map_`, since there's no point in holding on to refs to any cluster until this call completes.  So I think we don't need this block at all.

Reply:
Done, in the filter code.

Reply:
We still don't need this `if` block.  If this block wasn't here, then we'd return null anyway on line 1228.